UnitTests/tInductionPreprocessor.cpp

```diff
@@ -0,0 +1,359 @@
+  auto templ = env.signature->getFnDefHandler()->getInductionTemplate(p.functor(), false);
+  auto b = templ.branches();
+  ASS_EQ(b.size(), v.size());
+  TermList t;
+  for (unsigned i = 0; i < b.size(); i++) {
+    ASS_EQ(b[i]._header, v[i].first);
+    auto r = b[i]._recursiveCalls;
+    ASS_EQ(r.size(), v[i].second.size());
+    for (unsigned j = 0; j < r.size(); j++) {
+      ASS_EQ(r[j], v[i].second[j]);
+    }
+  }
+}
+
+inline void checkTemplateBranches(const FuncSugar& f, const vvector<pair<TermSugar, vvector<TermSugar>>>& p) {
+  ASS(env.signature->getFnDefHandler()->hasInductionTemplate(f.functor(), true));
+  auto templ = env.signature->getFnDefHandler()->getInductionTemplate(f.functor(), true);
+  auto b = templ.branches();
+  ASS_EQ(b.size(), p.size());
+  TermList t;
+  for (unsigned i = 0; i < b.size(); i++) {
+    ASS_EQ(b[i]._header, p[i].first.toTerm().term());
+    auto r = b[i]._recursiveCalls;
+    ASS_EQ(r.size(), p[i].second.size());
+    for (unsigned j = 0; j < r.size(); j++) {
+      ASS_EQ(r[j], p[i].second[j].toTerm().term());
+    }
+  }
+}
+
+
+inline void checkInductionTerms(const InductionScheme& sch, const InductionTerms p) {
+  ASS(sch.inductionTerms() == p);
+}
+
+inline void checkCases(const InductionScheme& sch, const vvector<pair<vmap<unsigned, TermList>, vvector<vmap<unsigned, TermList>>>>& p) {
+  auto c = sch.cases();
+  ASS_EQ(c.size(), p.size());
+  TermList t;
+  for (unsigned i = 0; i < c.size(); i++) {
+    for (const auto& kv : p[i].first) {
+      ASS(c[i]._step.findBinding(kv.first, t));
+      ASS_EQ(t, kv.second);
+    }
+    auto r = c[i]._recursiveCalls;
+    ASS_EQ(r.size(), p[i].second.size());
+    for (unsigned j = 0; j < r.size(); j++) {
+      for (const auto& kv : p[i].second[j]) {
+        ASS(r[j].findBinding(kv.first, t));
+        ASS_EQ(t, kv.second);
+      }
+    }
+  }
+}
+
+// not well-founded functions
+TEST_FUN(test_01) {
+  __ALLOW_UNUSED(MY_SYNTAX_SUGAR)
+  DECL_FUNC_DEFS({ { clause({ f(r(x), r(y)) == f(f(x, r(r(y))), y) }),          0, false },   \
+                   { clause({ g(r(x)) == g(f(x,x)) }),                          0, false },   \
+                   { clause({ h(x, y, r1(x, z)) == h(y, y, z) }),               0, false },   \
+                   { clause({ h(r(x), y, z) == h(x, x, r2(y,z)) }),             0, false },   \
+                   { clause({ ~p(r(r(x))), ~p(y) }),                            0, false },   \
+                   { clause({ q(r1(x,y),r(z)), ~q(y,r(z)), g(b) == b, q(z,b) }),0, false } })
+
+  ASS(!env.signature->getFnDefHandler()->hasInductionTemplate(f.functor(), true));
+  ASS(!env.signature->getFnDefHandler()->hasInductionTemplate(g.functor(), true));
+  ASS(!env.signature->getFnDefHandler()->hasInductionTemplate(h.functor(), true));
+  ASS(!env.signature->getFnDefHandler()->hasInductionTemplate(p.functor(), false));
+  ASS(!env.signature->getFnDefHandler()->hasInductionTemplate(q.functor(), false));
+}
+
+// not useful functions
```

Comment:
It means they are not recursive. I'll clarify this as well.

Reply:
Or they are recursive but their arguments don't change. This is kind of confused in the code as well. If there are no inductive arguments, it can simply mean there are no recursive calls or that none of them change in any recursive call. I'll add comments to clarify.